### ISSUE TITLE
Dynamically update a ConfigurationSet

### ISSUE THREAD

**Issue by groud** [tr11/python-configuration#9]
Hi, I just found out your library and I am quite thankful for sharing this work :)

I wanted to use python-configuration in my project, however I need the possibility to dynamically update a ConfigurationSet with new value (this is supposed to be exposed as a web service in my app).

Do you think this could be possible to implement ? I can code it myself if needed, but I would like to know how you would see the API.

Thanks !

**Comment by tr11** [comment]
Could you give an example of the type of update you had in mind?  Adding another configuration source or replacing a particular item in a `Configuration` instance?

**Comment by groud** [comment]
Here is my use case. I need to:
- load a configuration from a file
- receive regular updates to update the loaded configuration. The update contains a limited set of key/values to update.

So to solve that I would create two Configuration object, one built from the file, and one for the updates. Both need to be in a single ConfigurationSet, but the problem is that the second one needs to be updated.

I guess the solution could either be:
1) Allow to dynamically add and remove a configuration from a ConfigurationSet, so that I can recreate a configuration everytime I receive an update
2) Allow the modification of a created configuration, likely in the same way a dictionary would work (either by setting individual fields or with something like an update(otherConf) function)

IMHO, the second solution is the best one, since it avoids recreating a configuration object.

**Comment by tr11** [comment]
I also like 2 better. Maybe this could be solved by making the configuration objects optionally lazy or at least give the user the ability of forcing a refresh. We could store a reference to the original target (dict, path, etc) and (a) rebuild the configuration objects on access if in lazy mode, or (b) give access to underlying configurations of a configuration set and allow for manual refreshes.

**Comment by groud** [comment]
I don't think storing a reference is really necessary in fact, as if you want to keep the configuration data from the original source, you can simply instead use a ConfigurationSet, with two configurations objects: one for the file and another one you can modify. 

So I would add to the api:
- an accessor to the fields of the `Configuration` objects. Via a `__setattr__(...)` and a `set(...)` functions. Eventually a clear() function too.
- an `update(otherConfigurationObject)` function, that would update a Configuration object using another one (so that you can build this Configuration object thanks to any source you want)

That would make the Configuration object modifiable as the user needs, while still keeping the ConfigurationSet as the way to mix several Configuration together, while not modifying the underlying objects.

**Comment by tr11** [comment]
That makes sense. I assumed you wanted to reload the file periodically.

Is it correct to say that your proposed changes essentially bring some of the missing `dict` methods to make a `Configuration` object more dict-like?

**Comment by groud** [comment]
> That makes sense. I assumed you wanted to reload the file periodically.

Ah no, not the file, the file would be loaded only once.

> Is it correct to say that your proposed changes essentially bring some of the missing dict methods to make a Configuration object more dict-like?

Yes, exactly.

**Comment by tr11** [comment]
I'm making some changes based on this issue on this branch https://github.com/tr11/python-configuration/tree/add-dict-methods.  I added the missing methods to the `Configuration` class but need to think about what to do for `ConfigurationSet` instances, in particular when we want to update values.

**Comment by groud** [comment]
The changes in the branch look good to me. Though, I think making the `ConfigurationSet` writable doesn't really make sense though. I think it should be kept as a "proxy" to a list of `Configuration` objects, only the list of those `Configuration` objects should be modifiable (by adding, removing, clearing, etc... Configurations items to the list). A simple solution might be to make the list of configuration objects public, that would solve the problem easily.

**Comment by tr11** [comment]
I was debating the merits of both approaches.  At first, I implemented what you suggest but since a `ConfigurationSet` is also a `Configuration` object, it would be a bit off if it behaved differently than the main `Configuration` class.

**Comment by groud** [comment]
> ConfigurationSet is also a Configuration object

Ah, is it necessary ? I don't think ConfigurationSet is the same as a Configuration object. IMHO, I would not make ConfigurationSet inherit from Configuration, even if it would share part of the interface (to access underlying properties of the owned Configuration).

**Comment by tr11** [comment]
The idea of the `ConfigurationSet` was to be indistinguishable from a `Configuration` so we can freely move from using a configuration to using a set of configurations.  This arose from the need to override parameters in some test environments without making changes to the core configuration.
That said, I'm not completely sure about a writable `ConfigurationSet` class.  Maybe it would be better to refactor part of the project and add mixin classes that specifically indicate whether the configuration is writable (for example, conf sets, files, remote services such as azure are not writable).

**Comment by groud** [comment]
>  This arose from the need to override parameters in some test environments without making changes to the core configuration.

That's why the changes to `Configuration` solve the problem quite well, as you can load a set of "read only" configurations and mix them with a writable configuration in a `ConfigurationSet`.

Here is an example of the usage:
```
my_file_config  = config_from_json(...)
my_env_config  = config_from_env(...)
my_custom_config = Configuration()
my_custom_config.prop = 2
final_configuration = ConfigurationSet(my_file_config, my_env_config, my_custom_config)

# Even after the ConfigurationSet creation, you can still update the custom config, and it should be available via final_configuration
my_custom_config.other_prop = 2
```

To add to that, I would not bother implementing "read only" configurations. Python's philosophy is usually to let people do what they want as grown ups :). So if they want to modify a configuration loaded via a file (or any other source), I think it is ok to allow that.

**Comment by tr11** [comment]
I agree, but making the ConfigurationSet writable simplifies your example a bit *if* the custom configuration is the leading one:

```
my_file_config  = config_from_json(...)
my_env_config  = config_from_env(...)
final_configuration = ConfigurationSet(my_file_config, my_env_config)

# Even after the ConfigurationSet creation, you can still update the custom config, and it should be available via final_configuration
final_configuration.other_prop = 2
```

It may not be very useful on your example, but writable sets would probably be useful for the following: 
```
final_config = config('env', '/path/to/file.json')
final_config.update({'prop': 2})
```
which allows updates without being too verbose about creating sets.

**Comment by groud** [comment]
Ah right, I did not see that part of the API...

But in that case, why not make `config('env', '/path/to/file.json')` return a `Configuration` object instead ?
That would basically disallow you to go back to the behind-the-scene several `Configuration` objects, but with this method you already cannot do that.

**Comment by tr11** [comment]
The reason to keep `config('env', '/path/to/file.json')` as a `ConfigurationSet` was to implement a `reload` method in the future., meaning that we could ask the `ConfigurationSet` to reread the env variables and JSON file. This was just a plan and is not implemented yet. Incidentally, this is the feature I thought you referred to at the beginning of the thread.

**Comment by groud** [comment]
> Incidentally, this is the feature I thought you referred to at the beginning of the thread.

Ah no, not really. ^^

In fact, I don't think reloading is that important, as I don't think it is a really common use case.
Reloading a configuration could be made instead in a per-source basis, as I don't think you will need often to reload all sources. Consequently, I think simply allowing to add/delete/replace the Configurations in a ConfigurationSet would be enough for this use case.

I mean, here is how I see things: 
- Configuration is the "writable" part of the API: you can do pretty much what you want with it such as populating it with a source file, edit its entries etc...
- ConfigurationSet is just a proxy class to several Configuration objects, only allowing you to read the data from several objects with a simple API.

**Comment by tr11** [comment]
While I agree that is works for your use case, on all the projects we've been using this library we go with the simplified version
```
config('env', '/path/to/file1.json', '/path/to/file2.json')
```
which makes it reasonable to at least consider the resulting instance writable.

We also reload (by replacing an underlying configuration manually) on a per-source basis too, so probably not too important to reload the whole `ConfigurationSet`, especially now that remote sources were introduced (and are expensive to update).

I'll create another issue so we make the underlying configuration sources public.

**Comment by groud** [comment]
Well, the thing is that, in fact, I think I would prefer the ConfigurationSet not to be writable. It would avoid errors later on as you are kind of not supposed to be modifying the configuration you use once it is loaded. The problem I see is that making everything writable kind of limit the benefit of using ConfigurationSet compared to updating a Configuration using another.

But anyway, I don't mind that much. Provided it is possible to update a configuration file I am fine :)

**Comment by tr11** [comment]
Closed by c7a202749de4895f7e50e347e74837903b7ed3e0.
